cmd/csi-api-gen/generators/api.go

```diff
@@ -0,0 +1,432 @@
+	pkgs := generator.Packages{
+		&generator.DefaultPackage{
+			PackageName: snakeCaseToPackageName(group.name),
+			PackagePath: group.serverPkg(),
+			HeaderText:  []byte(headerComment),
+
+			GeneratorList: []generator.Generator{
+				&apiGroupGeneratedGenerator{
+					DefaultGen: generator.DefaultGen{
+						OptionalName: "api_group_generated",
+					},
+					groupDefinition: group,
+				},
+			},
+		},
+
+		&generator.DefaultPackage{
+			PackageName: "internal",
+			PackagePath: group.internalServerPkg(),
+			HeaderText:  []byte(headerComment),
+
+			GeneratorList: []generator.Generator{
+				&typesGeneratedGenerator{
+					DefaultGen: generator.DefaultGen{
+						OptionalName: "types_generated",
+					},
+					groupDefinition: group,
+				},
+			},
+		},
+	}
+
+	// generate a skeleton server.go if it doesn't exist
+	if !fileExists(path.Join(outputBase, group.serverPkg(), "server.go")) {
+		pkgs = append(pkgs, &generator.DefaultPackage{
+			PackageName: snakeCaseToPackageName(group.name),
+			PackagePath: group.serverPkg(),
+
+			GeneratorList: []generator.Generator{
+				&serverGenerator{
+					DefaultGen: generator.DefaultGen{
+						OptionalName: "server",
+					},
+					groupDefinition: group,
+				},
+			},
+		})
+	}
+
+	// generate types.go if it doesn't exist and the group only has one version
+	if len(group.versions) == 1 && !fileExists(path.Join(outputBase, group.internalServerPkg(), "types.go")) {
+		pkgs = append(pkgs, &generator.DefaultPackage{
+			PackageName: "internal",
+			PackagePath: group.internalServerPkg(),
+
+			GeneratorList: []generator.Generator{
+				&typesGenerator{
+					DefaultGen: generator.DefaultGen{
+						OptionalName: "types",
+					},
+					groupDefinition: group,
+					version:         group.versions[0],
+				},
+			},
+		})
+	}
+
+	for _, version := range group.versions {
+		// go closures...
+		vsn := version
+		pkgs = append(pkgs,
+			&generator.DefaultPackage{
+				PackageName: version.Name,
+				PackagePath: group.versionedServerPkg(version.Name),
+				HeaderText:  []byte(headerComment),
+
+				GeneratorFunc: func(context *generator.Context) []generator.Generator {
+					// this is somewhat of a hack: the conversion generator expects all inputs
+					// to be valid go packages - in particular, to have at least one valid go file;
+					// which does not hold true for new API versions
+					// so here we create a temporary, empty doc.go file that we clean up right after
+					docFilePath := path.Join(outputBase, group.versionedServerPkg(vsn.Name), "doc.go")
+					if !fileExists(docFilePath) {
+						if err := os.MkdirAll(path.Dir(docFilePath), os.ModePerm); err != nil {
+							klog.Fatalf("Unable to create directory %q: %v", path.Dir(docFilePath), err)
+						}
+						contents := fmt.Sprintf("// Temporary code generated by conversion-gen. DO NOT COMMIT.\n// Should be removed automatically, but please remove manually otherwise.\n\npackage %s\n", vsn.Name)
```

Comment:
Should this say `conversion-gen` or `csi-proxy-api-gen`?

Reply:
👍